util/limiter.py

```diff
@@ -15,54 +29,59 @@ def limit(_limit_level: int) -> callable:
     Levels are inclusive, so if you are level 2, you cannot use level 1 commands either.
     """
 
-    def decorator(func: callable) -> callable:
+    def decorator(
+        func: Callable[LimitedCommandParams, Awaitable[LimitedCommandReturnValue]],
+    ) -> Callable[LimitedCommandParams, Awaitable[LimitedCommandReturnValue | None]]:
         @functools.wraps(func)
-        async def wrapper(*args, **kwargs) -> callable:
+        async def wrapper(
+            *args: LimitedCommandParams.args,
+            **kwargs: LimitedCommandParams.kwargs,
+        ) -> LimitedCommandReturnValue | None:
             """
             Wrapper to determine if the user is limited or not.
             """
 
             # Get the author id
-            author_id = None
             for arg in args:
-                try:
-                    author_id = arg.author.id
+                if isinstance(arg, ApplicationContext):
                     ctx = arg
-                except AttributeError:  # Type is not a context
-                    continue
+                    break
+            else:
+                log(
+                    f"@limit() decorator was applied to non-slash-command {func!r}! "
+                    "Permitting function call without checking permissions!",
+                )
+                return await func(*args, **kwargs)
```

Comment:
As nice as this is, newer developers are not going to have the slightest idea how this works. I recommend simplifying this to:
```py
            ctx = None
            for arg in args:
                if isinstance(arg, ApplicationContext):
                    ctx = arg
                    break
            if ctx is None:
                log(
                    f"@limit() decorator was applied to non-slash-command {func!r}! "
                    "Permitting function call without checking permissions!",
                )
                return await func(*args, **kwargs)
```

Reply:
That's unfortunately true. For-else isn't very popular. Updated in 497d546b89539f5ca38bd8323377b82c18558e37